### PR TITLE
First version of Operations Events model

### PR DESCRIPTION
**Pull request by raisoman**

### CODE REVIEW

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 1]
```diff
@@ -515,12 +515,19 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
+
 DROP TABLE IF EXISTS dcsa_im_v3_0.event CASCADE;
 CREATE TABLE dcsa_im_v3_0.event (
-    event_id uuid DEFAULT uuid_generate_v4() PRIMARY KEY,
+    event_id uuid DEFAULT uuid_generate_v4() PRIMARY KEY, -- Unique identifier for the event captured.
+    event_created_date_time timestamp with time zone NOT NULL DEFAULT now(), -- The date and time when the event record was created and stored.
     event_type text NOT NULL,
-    event_classifier_code varchar(3) NOT NULL,
-    event_date_time timestamp with time zone NOT NULL,
+    event_classifier_code varchar(3) NOT NULL, -- Code for the event classifier telling whether the information relates to an actual or future event.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:529]:
```suggestion
    event_classifier_code varchar(3) NOT NULL REFERENCES dcsa_im_v3_0.event_classifier(event_classifier_code), -- Code for the event classifier telling whether the information relates to an actual or future event.
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 2]
```diff
@@ -515,12 +515,19 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
+    event_date_time timestamp with time zone NOT NULL, -- Indicating the date and time of when the event occurred or will occur.
     event_type_code varchar(4) NOT NULL
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:531]: Didn't we agree to deprecate this and used `eventSubtype` instead

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 3]
```diff
@@ -544,6 +551,45 @@ CREATE TABLE dcsa_im_v3_0.transport_event (
     transport_call_id uuid NOT NULL
 ) INHERITS (dcsa_im_v3_0.event);
 
+DROP TABLE IF EXISTS dcsa_im_v3_0.operations_event CASCADE;
+CREATE TABLE dcsa_im_v3_0.operations_event (
+    operations_event_type_code varchar(4) NOT NULL, -- The code to identify the type of event that is related to the operation.
+    event_location uuid NOT NULL, -- The location where the event takes place.
+    port_call_service_type_code varchar(4) NOT NULL, -- The type of the service provided in the port call.
+    facility_type_code varchar(4) NULL -- Four character code to identify the specific type of facility.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:559]: Make a REFERENCES on this

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 4]
```diff
@@ -515,12 +515,19 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
 
 /* Events related Entities */
 
+DROP TABLE IF EXISTS dcsa_im_v3_0.event_classifier CASCADE;
+CREATE TABLE dcsa_im_v3_0.event_classifier (
+    event_classifier_code char(3) PRIMARY KEY, -- Code for the event classifier, either PLN, ACT or EST.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:520]:
```suggestion
    event_classifier_code char(3) PRIMARY KEY, -- Code for the event classifier, either PLN, ACT, EST, REQ or TPL.
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 5]
```diff
@@ -515,35 +515,117 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
+DROP TABLE IF EXISTS dcsa_im_v3_0.operations_event_type CASCADE;
+CREATE TABLE dcsa_im_v3_0.operations_event_type (
+    operations_event_type_code varchar(4) PRIMARY KEY,
+    operations_event_type_name varchar(30) NOT NULL,
+    operations_event_type_description varchar(200) NOT NULL
+);
+
+ALTER TABLE dcsa_im_v3_0.operations_event
+ADD FOREIGN KEY (event_classifier_code)
+REFERENCES dcsa_im_v3_0.event_classifier (event_classifier_code);
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:595]: Is this a good way to do it? My impression was that ShipmentEvents and TransportEvents can have different eventClassifierCodes.
Wouldn't a better way to add a CONSTRAINT in each of the event classes to only allow specific values?

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 6]
```diff
@@ -532,35 +532,35 @@ CREATE TABLE dcsa_im_v3_0.event (
...
 
 DROP TABLE IF EXISTS dcsa_im_v3_0.operations_event CASCADE;
 CREATE TABLE dcsa_im_v3_0.operations_event (
-    operations_event_type_code varchar(4) NOT NULL, -- The code to identify the type of event that is related to the operation.
-    event_location uuid NOT NULL, -- The location where the event takes place.
-    port_call_service_type_code varchar(4) NOT NULL, -- The type of the service provided in the port call.
+    operations_event_type_code varchar(4) NOT NULL REFERENCES dcsa_im_v3_0.operations_event_type(operations_event_type_code), -- The code to identify the type of event that is related to the operation.
+    event_location uuid NOT NULL REFERENCES dcsa_im_v3_0.location(id), -- The location where the event takes place.
+    port_call_service_type_code varchar(4) NOT NULL REFERENCES dcsa_im_v3_0.port_call_service_type(port_call_service_type_code), -- The type of the service provided in the port call.
     facility_type_code varchar(4) NULL -- Four character code to identify the specific type of facility.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:564]:
```suggestion
    facility_type_code varchar(4) NULL REFERENCES dcsa_im_v3_0.facility_type(facility_type_code) -- Four character code to identify the specific type of facility.
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 7]
```diff
@@ -515,35 +427,60 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
-    event_type_code varchar(4) NOT NULL
+    event_classifier_code varchar(3) NOT NULL, -- Code for the event classifier telling whether the information relates to an actual or future event.
+    event_date_time timestamp with time zone NOT NULL -- Indicating the date and time of when the event occurred or will occur.
 );
 
 DROP TABLE IF EXISTS dcsa_im_v3_0.equipment_event CASCADE;
 CREATE TABLE dcsa_im_v3_0.equipment_event (
+    equipment_event_type_code varchar(4) NOT NULL REFERENCES dcsa_im_v3_0.equipment_event_type(equipment_event_type_code),
     equipment_reference varchar(15),
     empty_indicator_code text NOT NULL,
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:449]:
```suggestion
    empty_indicator_code text(5) NOT NULL,
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 8]
```diff
@@ -515,35 +427,60 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
     vessel_schedule_change_remark varchar(250),
+    location_id uuid REFERENCES dcsa_im_v3_0.location(id),
+    facility_type_code varchar(4) REFERENCES dcsa_im_v3_0.facility_type(facility_type_code),
     transport_call_id uuid NOT NULL
 ) INHERITS (dcsa_im_v3_0.event);
 
+DROP TABLE IF EXISTS dcsa_im_v3_0.operations_event CASCADE;
+CREATE TABLE dcsa_im_v3_0.operations_event (
+    operations_event_type_code varchar(4) NOT NULL REFERENCES dcsa_im_v3_0.operations_event_type(operations_event_type_code), -- The code to identify the type of event that is related to the operation.
+    event_location uuid NOT NULL REFERENCES dcsa_im_v3_0.location(id), -- The location where the event takes place.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:474]:
```suggestion
    publisher varchar(50) NOT NULL, -- The publisher (source) of the event
    publisher_role varchar(3) NOT NULL, -- The party function code of the publisher
    event_location uuid NOT NULL REFERENCES dcsa_im_v3_0.location(id), -- The location where the event takes place.
```

**Comment by HenrikHL** [comment]
Delay reason code table is missing

File: datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+DROP TABLE IF EXISTS dcsa_im_v3_0.country CASCADE;
+CREATE TABLE dcsa_im_v3_0.country (
+    country_code varchar(2) PRIMARY KEY,
+    country_name varchar(75) NULL
+);
+
+DROP TABLE IF EXISTS dcsa_im_v3_0.equipment_event_type CASCADE;
+CREATE TABLE dcsa_im_v3_0.equipment_event_type (
+    equipment_event_type_code varchar(4) PRIMARY KEY,
+    equipment_event_type_name varchar(30) NOT NULL
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql:17]: description is missing

File: datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    mode_of_transport_code varchar(3) PRIMARY KEY,
+    mode_of_transport_name varchar(100) NULL,
+    mode_of_transport_description varchar(250) NULL,
+    dcsa_transport_type varchar(50) NULL UNIQUE
+);
+
+DROP TABLE IF EXISTS dcsa_im_v3_0.operations_event_type CASCADE;
+CREATE TABLE dcsa_im_v3_0.operations_event_type (
+    operations_event_type_code varchar(4) PRIMARY KEY,
+    operations_event_type_name varchar(30) NOT NULL
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql:52]: description is missing

File: datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+CREATE TABLE dcsa_im_v3_0.service_type (
+    service_type varchar(3) PRIMARY KEY,
+    description varchar(300) NOT NULL
+);
+
+DROP TABLE IF EXISTS dcsa_im_v3_0.shipment_event_type CASCADE;
+CREATE TABLE dcsa_im_v3_0.shipment_event_type (
+    shipment_event_type_code varchar(4) PRIMARY KEY,
+    shipment_event_type_name varchar(30) NOT NULL,
+    shipment_event_type_description varchar(200) NOT NULL
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql:98]:
```suggestion
    shipment_event_type_description varchar(250) NOT NULL
```

File: datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+CREATE TABLE dcsa_im_v3_0.transport_document_type (
+    transport_document_type_code varchar(3) PRIMARY KEY,
+    transport_document_type_name varchar(20) NULL,
+    transport_document_type_description varchar(500) NULL
+);
+
+DROP TABLE IF EXISTS dcsa_im_v3_0.transport_event_type CASCADE;
+CREATE TABLE dcsa_im_v3_0.transport_event_type (
+    transport_event_type_code varchar(4) PRIMARY KEY,
+    transport_event_type_name varchar(30) NOT NULL
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_1_dcsa_im_v3_0_reference_data_tables.sql:117]: description is missing

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 13]
```diff
@@ -515,35 +427,60 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
 
 /* Events related Entities */
 
+DROP TABLE IF EXISTS dcsa_im_v3_0.event_classifier CASCADE;
+CREATE TABLE dcsa_im_v3_0.event_classifier (
+    event_classifier_code char(3) PRIMARY KEY, -- Code for the event classifier, either PLN, ACT or EST.
+    event_classifier_name varchar(100) NULL -- Name of the classifier.
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:433]:
```suggestion
    event_classifier_name varchar(30) NULL, -- Name of the classifier.
    event_classifier_description varchar(250) NULL -- The description of the event classifier. 
```

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 14]
```diff
@@ -515,35 +427,60 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
+    shipment_information_type_code varchar(3) NOT NULL,
+    document_id varchar(50) NOT NULL,
+    reason varchar(100)
 ) INHERITS (dcsa_im_v3_0.event);
 
 DROP TABLE IF EXISTS dcsa_im_v3_0.transport_event CASCADE;
 CREATE TABLE dcsa_im_v3_0.transport_event (
+    transport_event_type_code varchar(4) NOT NULL REFERENCES dcsa_im_v3_0.transport_event_type(transport_event_type_code),
     delay_reason_code varchar(3),
     vessel_schedule_change_remark varchar(250),
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:465]:
```suggestion
    change_remark varchar(250), -- Free text description of the reason for the change in schedule. 
```

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:465]: to align with IM

File: datamodel/initdb.d/03_dcsa_im_v3_0.sql [hunk 15]
```diff
@@ -515,35 +427,60 @@ CREATE TABLE dcsa_im_v3_0.shipment_transport (
...
+    location_id uuid REFERENCES dcsa_im_v3_0.location(id),
+    facility_type_code varchar(4) REFERENCES dcsa_im_v3_0.facility_type(facility_type_code),
```
Comment (HenrikHL) [inline datamodel/initdb.d/03_dcsa_im_v3_0.sql:467]:
```suggestion
``

Comment (HenrikHL) [reply datamodel/initdb.d/03_dcsa_im_v3_0.sql:467]: delete this to align with IM

**Review by HenrikHL — APPROVED** [review]

**Comment by douradofrazer** [comment]
As discussed with @raisoman this PR is obsolete and will be reworked with future PRs.

**Comment by douradofrazer** [comment]
Closing this PR as its obsolete now, as discussed with @raisoman .

**Pull request closed without merging**